File: java/lasagna/src/main/java/Lasagna.java
public class Lasagna {
    /**
    * Returns how many minutes the lasagna should be in the oven.
    *
    * @return      the expected oven time
    */
    public int expectedMinutesInOven() {
        return 40;
    }

    /**
    * Returns how many minutes the lasagna still has to remain in the oven, 
    * based on the expected oven time in minutes
    *
    * @param  minutesInOven  the actual minutes the lasagna has been in the oven
    * @return      remaining minutes in oven
    */
    public int remainingMinutesInOven(int minutesInOven) {
        return expectedMinutesInOven() - minutesInOven;
    }

    /**
    * Returns how many minutes you spent preparing the lasagna, assuming 
    * each layer takes you 2 minutes to prepare.
    *
    * @param  layers  the number of layers
    * @return      time it takes to prepare the lasagna
    */
    public int preparationTimeInMinutes(int layers) {
        return layers * minutesPerLayer();
    }

    /**
    * Returns how many minutes in total you've worked on cooking the lasagna.
    *
    * @param  layers  the number of layers
    * @param  minutesInOven  the actual minutes the lasagna has been in the oven
    * @return      time it takes to prepare the lasagna
    */
    public int totalTimeInMinutes(int layers, int minutesInOven) {
        return preparationTimeInMinutes(layers) + minutesInOven;
    }

    private int minutesPerLayer() {
        return 2;
    }
}

File: elixir/lasagna/lib/lasagna.ex
defmodule Lasagna do
  @spec expected_minutes_in_oven() :: number()
  def expected_minutes_in_oven do
    40
  end

  @spec remaining_minutes_in_oven(number()) :: number()
  def remaining_minutes_in_oven(minutes_in_oven) do
    expected_minutes_in_oven() - minutes_in_oven
  end

  @spec preparation_time_in_minutes(number()) :: number()
  def preparation_time_in_minutes(layers) do
    2 * layers
  end

  @spec total_time_in_minutes(number(), number()) :: number()
  def total_time_in_minutes(layers, minutes_in_oven) do
    preparation_time_in_minutes(layers) + minutes_in_oven
  end

  @spec alarm() :: String.t()
  def alarm do
    "Ding!"
  end
end
